scripts/msct_moco.py

```diff
@@ -66,160 +68,242 @@ def moco(param):
+    # file_mat = tuple([[[] for i in range(nt)] for i in range(nz)])
+    file_mat[:] = ''  # init
+    file_data_splitZ_moco = []
+    sct.printv('\nRegister. Loop across Z (note: there is only one Z if orientation is axial')
+    for file in file_data_splitZ:
+        iz = file_data_splitZ.index(file)
+        # Split data along T dimension
+        # sct.printv('\nSplit data along T dimension.', verbose)
+        im_z = Image(file)
+        list_im_zt = split_data(im_z, dim=3)
+        file_data_splitZ_splitT = []
+        for im_zt in list_im_zt:
+            im_zt.save(verbose=0)
+            file_data_splitZ_splitT.append(im_zt.absolutepath)
+        # file_data_splitT = file_data + '_T'
+
+        # Motion correction: initialization
+        index = np.arange(nt)
+        file_data_splitT_num = []
+        file_data_splitZ_splitT_moco = []
+        failed_transfo = [0 for i in range(nt)]
+
+        # Motion correction: Loop across T
+        tqdm_bar = tqdm(total=nt, unit='iter', unit_scale=False,
+                        desc="Z=" + str(iz) + "/" + str(len(file_data_splitZ)-1), ascii=True, ncols=80)
+        for indice_index in range(nt):
+
+            # create indices and display stuff
+            it = index[indice_index]
+            file_mat[iz][it] = os.path.join(folder_mat, "mat.Z") + str(iz).zfill(4) + 'T' + str(it).zfill(4)
+            file_data_splitZ_splitT_moco.append(sct.add_suffix(file_data_splitZ_splitT[it], '_moco'))
+            # deal with masking
+            if not param.fname_mask == '':
+                input_mask = im_maskz_list[iz]
+            else:
+                input_mask = None
+            # run 3D registration
+            failed_transfo[it] = register(param, file_data_splitZ_splitT[it], file_target_splitZ[iz], file_mat[iz][it],
+                                          file_data_splitZ_splitT_moco[it], im_mask=input_mask)
+
+            # average registered volume with target image
+            # N.B. use weighted averaging: (target * nb_it + moco) / (nb_it + 1)
+            if param.iterAvg and indice_index < 10 and failed_transfo[it] == 0 and not param.todo == 'apply':
+                im_targetz = Image(file_target_splitZ[iz])
+                data_targetz = im_targetz.data
+                data_mocoz = Image(file_data_splitZ_splitT_moco[it]).data
+                data_targetz = (data_targetz * (indice_index + 1) + data_mocoz) / (indice_index + 2)
+                im_targetz.data = data_targetz
+                im_targetz.save(verbose=0)
+            tqdm_bar.update()
+        tqdm_bar.close()
+
+        # Replace failed transformation with the closest good one
+        fT = [i for i, j in enumerate(failed_transfo) if j == 1]
+        gT = [i for i, j in enumerate(failed_transfo) if j == 0]
+        for it in range(len(fT)):
+            abs_dist = [abs(gT[i] - fT[it]) for i in range(len(gT))]
```

Comment:
Use numpy?

Reply:
fixed in 7b3647c678dc874a9146d6cdc34cf950d1596dbd